interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,351 @@
+func (pdb *PromptDB) AddOrMerge(user uint32, snap string, iface string, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, exists := pdb.perUser[user]
+	if !exists {
+		pdb.perUser[user] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[user]
+	}
+
+	availablePermissions, _ := prompting.AvailablePermissions(iface)
+	// Error should be impossible, since caller has already validated that iface
+	// is valid, and tests check that all valid interfaces have valid available
+	// permissions returned by AvailablePermissions.
+
+	constraints := &promptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == snap && prompt.Interface == iface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			return prompt, true
+		}
+	}
+
+	id := pdb.nextID()
+	timestamp := time.Now()
+	prompt := &Prompt{
+		ID:           id,
+		Timestamp:    timestamp,
+		Snap:         snap,
+		Interface:    iface,
+		Constraints:  constraints,
+		listenerReqs: []*listener.Request{listenerReq},
+	}
+	userEntry.ByID[id] = prompt
+	pdb.notifyPrompt(user, id)
+	return prompt, false
+}
+
+// Prompts returns a slice of all outstanding prompts.
+func (pdb *PromptDB) Prompts(user uint32) []*Prompt {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, exists := pdb.perUser[user]
+	if !exists {
+		return make([]*Prompt, 0)
+	}
+	prompts := make([]*Prompt, 0, len(userEntry.ByID))
+	for _, prompt := range userEntry.ByID {
+		prompts = append(prompts, prompt)
+	}
+	return prompts
+}
+
+// PromptWithID returns the prompt with the given ID for the given user.
+func (pdb *PromptDB) PromptWithID(user uint32, id string) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, exists := pdb.perUser[user]
+	if !exists {
+		return nil, fmt.Errorf("cannot find prompts for the given UID: %d", user)
+	}
+	prompt, exists := userEntry.ByID[id]
+	if !exists {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	return prompt, nil
+}
+
+// Reply resolves the prompt with the given ID using the given outcome by
+// sending a reply to all associated listener requests, then removing the
+// prompt from the prompt DB.
+//
+// Records a notice for the prompt, and returns the prompt's former contents.
+func (pdb *PromptDB) Reply(user uint32, id string, outcome prompting.OutcomeType) (*Prompt, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, exists := pdb.perUser[user]
+	if !exists || len(userEntry.ByID) == 0 {
+		return nil, fmt.Errorf("cannot find prompts for the given UID: %d", user)
```

Comment:
bit of a nitpick but an empty ByID is handled differently here and in the method above when otherwise it's the same

Reply:
I think when we expect to find a prompt with a particular ID but it doesn't exist, the error message should be the same regardless of whether any prompts exist for that user, so I'm adding the `len(userEntry.ByID) == 0` check to the other relevant places and made the error messages all be `fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)`.